bin/cheritest/cheritest_registers.c

```diff
@@ -148,7 +148,7 @@ check_initreg_code(void * __capability c)
 	 * length).
 	 */
 	vaddr_t upper_bound =
-	    CHERI_REPRESENTABLE_LENGTH(cheri_getaddress(c) + 0x100000);
+	    CHERI_REPRESENTABLE_LENGTH(cheri_getaddress(c) + 0x1000000);
```

Comment:
Seems like MIPS is not emitting a separate relro segment so the captable is closer to .text:
```
alr48@waltham:/local/scratch/alr48/cheri/output> $CHERI_SDK/llvm-readelf --program-headers rootfs-riscv64-purecap/bin/cheriabitest

Elf file type is EXEC (Executable file)
Entry point 0x5b9e0
There are 10 program headers, starting at offset 64

Program Headers:
  Type           Offset   VirtAddr           PhysAddr           FileSiz  MemSiz   Flg Align
  PHDR           0x000040 0x0000000000010040 0x0000000000010040 0x000230 0x000230 R   0x8
  LOAD           0x000000 0x0000000000010000 0x0000000000010000 0x04a9d8 0x04a9d8 R   0x1000
  LOAD           0x04a9e0 0x000000000005b9e0 0x000000000005b9e0 0x0b062c 0x0b062c R E 0x1000
  LOAD           0x0fc000 0x000000000010e000 0x000000000010e000 0x019a40 0x019a40 RW  0x1000
  LOAD           0x116a40 0x0000000000128a40 0x0000000000128a40 0x006724 0x4af4b4 RW  0x2000
  TLS            0x0fc000 0x000000000010e000 0x000000000010e000 0x0021a0 0x005020 R   0x1000
  GNU_RELRO      0x0fc000 0x000000000010e000 0x000000000010e000 0x019a40 0x01a000 R   0x1
  GNU_EH_FRAME   0x00f098 0x000000000001f098 0x000000000001f098 0x003854 0x003854 R   0x4
  GNU_STACK      0x000000 0x0000000000000000 0x0000000000000000 0x000000 0x000000 RW  0x0
  NOTE           0x000270 0x0000000000010270 0x0000000000010270 0x000048 0x000048 R   0x4

 Section to Segment mapping:
  Segment Sections...
   00
   01     .note.tag .rodata .eh_frame_hdr .eh_frame __cap_relocs
   02     .text
   03     .tdata .data.rel.ro .init_array .fini_array .ctors .dtors .captable
   04     .data .sdata .bss .sbss
   05     .tdata .tbss
   06     .tdata .data.rel.ro .init_array .fini_array .ctors .dtors .captable
   07     .eh_frame_hdr
   08
   09     .note.tag
   None   .comment .debug_loc .debug_abbrev .debug_info .debug_ranges .debug_str .riscv.attributes .debug_line .debug_aranges .gnu.warning.mktemp .gnu.warning.f_prealloc .symtab .shstrtab .strtab
alr48@waltham:/local/scratch/alr48/cheri/output> $CHERI_SDK/llvm-readelf --program-headers rootfs-purecap128/bin/cheriabitest

Elf file type is EXEC (Executable file)
Entry point 0x1200711d0
There are 10 program headers, starting at offset 64

Program Headers:
  Type           Offset   VirtAddr           PhysAddr           FileSiz  MemSiz   Flg Align
  PHDR           0x000040 0x0000000120000040 0x0000000120000040 0x000230 0x000230 R   0x8
  LOAD           0x000000 0x0000000120000000 0x0000000120000000 0x0611d0 0x0611d0 R   0x10000
  LOAD           0x0611d0 0x00000001200711d0 0x00000001200711d0 0x0eb5f0 0x0eb5f0 R E 0x10000
  LOAD           0x14d000 0x000000012016d000 0x000000012016d000 0x025bd0 0x4d2f00 RW  0x10000
  TLS            0x14d000 0x000000012016d000 0x000000012016d000 0x0021a0 0x005020 R   0x1000
  GNU_EH_FRAME   0x016234 0x0000000120016234 0x0000000120016234 0x003d84 0x003d84 R   0x4
  GNU_STACK      0x000000 0x0000000000000000 0x0000000000000000 0x000000 0x000000 RW  0x0
  NOTE           0x000270 0x0000000120000270 0x0000000120000270 0x000048 0x000048 R   0x4
  OPTIONS        0x0002d0 0x00000001200002d0 0x00000001200002d0 0x000028 0x000028 R   0x8
  ABIFLAGS       0x0002b8 0x00000001200002b8 0x00000001200002b8 0x000018 0x000018 R   0x8

 Section to Segment mapping:
  Segment Sections...
   00
   01     .note.tag .MIPS.abiflags .MIPS.options .rodata .eh_frame_hdr .eh_frame __cap_relocs
   02     .text
   03     .tdata .data .data.rel.ro .init_array .fini_array .ctors .dtors .captable .got .bss
   04     .tdata .tbss
   05     .eh_frame_hdr
   06
   07     .note.tag
   08     .MIPS.options
   09     .MIPS.abiflags
   None   .comment .pdr .debug_loc .debug_abbrev .debug_info .debug_ranges .debug_str .debug_line .debug_aranges .gnu.warning.mktemp .gnu.warning.f_prealloc .symtab .shstrtab .strtab
```

We could reduce the bounds of PCC if we always place .captable as close to .text as possible and add a new PT_CHERI_CAPTABLE (or something like that) so that we can find the end of the captable and bounds $pcc sensible in static binaries. For dynamic binaries RTLD knows where the captable ends since it gets that information from .dynamic.

Reply:
We should probably do a dl_iterate_phdrs to find a sensible upper bound rather than this hacky constant.